File: tests/test_l2_integration.py
"""
test_l2_integration.py - L2 Integration Tests
===============================================

Tests interactions between components.
Verifies that components work together correctly.
"""

import pytest
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent))

from everything_becomes_f.counter import UnsignedShortCounter
from everything_becomes_f.electromagnetic_lock import ElectromagneticLock, LockSystem
from everything_becomes_f.security_camera import SecurityCamera, ClockSyncVulnerability
from everything_becomes_f.sealed_room import SealedRoom, MagataQuarters, RoomState
from everything_becomes_f.event_system import EventBus, EventType
from everything_becomes_f.red_magic_system import RedMagicSystem, SystemState, create_standard_facility


# =============================================================================
# Counter + Lock Integration Tests
# =============================================================================

class TestCounterLockIntegration:
    """Tests counter overflow triggering lock release."""
    
    def test_overflow_triggers_lock_release(self):
        """Counter overflow should trigger lock failsafe."""
        counter = UnsignedShortCounter()
        lock_system = LockSystem("Test")
        lock_system.create_lock("lock_1", "Door 1")
        lock_system.create_lock("lock_2", "Door 2")
        
        # Register callback to trigger failsafe on overflow
        def on_overflow(c):
            lock_system.trigger_global_failsafe()
        
        counter.register_overflow_callback(on_overflow)
        
        # Verify locks are initially locked
        assert lock_system.all_locked()
        
        # Cause overflow
        counter.set_value(0xFFFF)
        counter.increment()
        
        # Verify locks are now unlocked
        assert lock_system.all_unlocked()
        assert lock_system.failsafe_triggered
    
    def test_fast_forward_to_overflow_unlocks(self):
        """Fast forward to overflow should unlock all locks."""
        counter = UnsignedShortCounter()
        lock_system = LockSystem("Test")
        lock_system.create_lock("lock_1", "Door 1")
        
        def on_overflow(c):
            lock_system.trigger_global_failsafe()
        
        counter.register_overflow_callback(on_overflow)
        
        # Fast forward past overflow
        counter.fast_forward(65536)
        
        assert lock_system.all_unlocked()


# =============================================================================
# Event Propagation Tests
# =============================================================================

class TestEventPropagation:
    """Tests event chain propagation through the system."""
    
    def test_overflow_event_chain(self):
        """Overflow should trigger event chain: overflow → locks → room."""
        event_bus = EventBus("Test")
        events_received = []
        
        # Track all events
        def handler(event):
            events_received.append(event.event_type)
        
        event_bus.subscribe_all(handler)
        
        # Simulate event chain
        event_bus.emit(EventType.COUNTER_OVERFLOW, source="counter")
        event_bus.emit(EventType.FAILSAFE_TRIGGERED, source="system")
        event_bus.emit(EventType.LOCKS_RELEASED, source="locks")
        event_bus.emit(EventType.ROOM_BREACHED, source="room")
        
        assert EventType.COUNTER_OVERFLOW in events_received
        assert EventType.LOCKS_RELEASED in events_received
        assert EventType.ROOM_BREACHED in events_received
        
        # Verify order
        assert events_received.index(EventType.COUNTER_OVERFLOW) < \
               events_received.index(EventType.LOCKS_RELEASED)
        assert events_received.index(EventType.LOCKS_RELEASED) < \
               events_received.index(EventType.ROOM_BREACHED)
    
    def test_event_bus_component_communication(self):
        """Components should communicate via event bus."""
        event_bus = EventBus("Test")
        
        # Counter component
        counter = UnsignedShortCounter()
        
        # Lock component - subscribes to overflow
        lock_system = LockSystem("Test")
        lock_system.create_lock("lock_1", "Door 1")
        
        def on_counter_overflow(event):
            lock_system.trigger_global_failsafe()
            event_bus.emit(EventType.LOCKS_RELEASED, source="lock_system")
        
        event_bus.subscribe(EventType.COUNTER_OVERFLOW, on_counter_overflow)
        
        # Room component - subscribes to locks released
        room = SealedRoom("room_1", "Test Room")
        
        def on_locks_released(event):
            room.breach()
            event_bus.emit(EventType.ROOM_BREACHED, source="room")
        
        event_bus.subscribe(EventType.LOCKS_RELEASED, on_locks_released)
        
        # Trigger the chain by emitting overflow
        event_bus.emit(EventType.COUNTER_OVERFLOW, source="counter")
        
        # Verify end state
        assert lock_system.all_unlocked()
        assert room.is_breached


# =============================================================================
# Clock Sync Vulnerability + Camera Tests
# =============================================================================

class TestClockSyncCameraIntegration:
    """Tests clock sync vulnerability creating camera blind spots."""
    
    def test_vulnerability_creates_blind_spot(self):
        """Exploiting vulnerability should create blind spots on cameras."""
        cameras = [
            SecurityCamera(camera_id="cam_1", location="Room 1"),
            SecurityCamera(camera_id="cam_2", location="Room 2"),
        ]
        
        # Record some footage first
        for cam in cameras:
            for minute in range(60):
                cam.record_minute(100, minute)
        
        vulnerability = ClockSyncVulnerability(cameras)
        
        # Exploit at specific time
        affected = vulnerability.exploit(100, 30)
        
        assert affected == 2
        
        # All cameras should have blind spot at that time
        for cam in cameras:
            assert cam.has_blind_spots()
            blind_spots = cam.get_blind_spots()
            assert any(bs.minute_index == 30 for bs in blind_spots)
    
    def test_reinstall_simulation(self):
        """Simulating reinstall should create blind spot."""
        cameras = [SecurityCamera(camera_id="cam_1", location="Room 1")]
        vulnerability = ClockSyncVulnerability(cameras)
        
        result = vulnerability.simulate_reinstall(50, 15)
        
        assert result["reinstall_triggered"]
        assert result["blind_spot_created"]
        assert result["clock_drift_minutes"] == 1
        assert result["blind_spot_time"] == (50, 15)


# =============================================================================
# Red Magic System Integration Tests
# =============================================================================

class TestRedMagicSystemIntegration:
    """Tests Red Magic system state transitions and component coordination."""
    
    def test_system_state_transitions(self, running_system):
        """System should transition through states correctly."""
        assert running_system.state == SystemState.RUNNING
        
        # Fast forward to overflow
        running_system.fast_forward_to_overflow()
        
        # Should be in SYSTEM_DOWN after overflow
        assert running_system.state == SystemState.SYSTEM_DOWN
    
    def test_system_initializing_to_running(self, initialized_system):
        """System should transition from INITIALIZING to RUNNING."""
        assert initialized_system.state == SystemState.INITIALIZING
        
        initialized_system.start()
        
        assert initialized_system.state == SystemState.RUNNING
    
    def test_overflow_triggers_full_sequence(self, running_system):
        """Overflow should trigger: OVERFLOW_DETECTED → FAILSAFE → SYSTEM_DOWN."""
        # Add a room to the system
        room = MagataQuarters()
        running_system.add_sealed_room(room)
        running_system.add_lock(room.main_lock)
        
        # Fast forward to overflow
        running_system.fast_forward_to_overflow()
        
        # Check final state
        assert running_system.state == SystemState.SYSTEM_DOWN
        assert running_system.lock_system.failsafe_triggered
        assert room.is_breached
    
    def test_system_events_during_overflow(self, running_system):
        """System should emit events during overflow sequence."""
        event_types_received = []
        
        def handler(event):
            event_types_received.append(event.event_type)
        
        running_system.event_bus.subscribe_all(handler)
        
        # Trigger overflow
        running_system.fast_forward_to_overflow()
        
        # Should have key events
        assert EventType.COUNTER_OVERFLOW in event_types_received
        assert EventType.FAILSAFE_TRIGGERED in event_types_received
        assert EventType.LOCKS_RELEASED in event_types_received
        assert EventType.SYSTEM_CRASH in event_types_received
    
    def test_standard_facility_setup(self, standard_facility):
        """Standard facility should have all components configured."""
        assert standard_facility.lock_system.lock_count >= 1
        assert len(standard_facility.cameras) >= 1
        assert len(standard_facility.sealed_rooms) >= 1
        assert standard_facility.state == SystemState.INITIALIZING


# =============================================================================
# Room + Lock Integration Tests
# =============================================================================

class TestRoomLockIntegration:
    """Tests sealed room behavior with electromagnetic locks."""
    
    def test_room_breach_when_locks_released(self, sealed_room):
        """Room should breach when its lock is released."""
        # Get the door's lock
        door = sealed_room.doors[0]
        
        # Release the lock
        door.lock.trigger_failsafe()
        
        # Room should still be sealed (lock release doesn't auto-breach)
        assert sealed_room.is_sealed
        
        # But now breach is possible
        sealed_room.breach()
        assert sealed_room.is_breached
    
    def test_magata_quarters_full_escape_sequence(self, magata_quarters):
        """Full escape sequence through Magata's quarters."""
        # Initial state
        assert magata_quarters.is_sealed
        assert magata_quarters.main_lock.is_locked()
        
        # Trigger overflow escape
        magata_quarters.trigger_overflow_escape(65536)
        
        # Final state
        assert magata_quarters.state == RoomState.ESCAPED
        assert magata_quarters.main_lock.is_unlocked()
        assert not magata_quarters.occupant_present


# =============================================================================
# System-Wide Event Tracking Tests
# =============================================================================

class TestSystemWideEvents:
    """Tests event tracking across entire system."""
    
    def test_complete_event_sequence(self, standard_facility):
        """Track complete event sequence from start to crash."""
        events = []
        
        def handler(event):
            events.append(event.event_type.name)
        
        standard_facility.event_bus.subscribe_all(handler)
        
        # Start system
        standard_facility.start()
        
        # Run to overflow
        standard_facility.fast_forward_to_overflow()
        
        # Should have complete sequence
        assert "SYSTEM_STARTED" in events
        assert "COUNTER_OVERFLOW" in events
        assert "FAILSAFE_TRIGGERED" in events
        assert "LOCKS_RELEASED" in events
        assert "SYSTEM_CRASH" in events
    
    def test_event_source_tracking(self, running_system):
        """Events should have correct source attribution."""
        events_by_source = {}
        
        def handler(event):
            source = event.source
            if source not in events_by_source:
                events_by_source[source] = []
            events_by_source[source].append(event.event_type)
        
        running_system.event_bus.subscribe_all(handler)
        running_system.fast_forward_to_overflow()
        
        # Should have events from multiple sources
        assert "UnsignedShortCounter" in events_by_source
        assert "RedMagicSystem" in events_by_source


# =============================================================================
# Run Tests
# =============================================================================

if __name__ == "__main__":
    pytest.main([__file__, "-v"])

File: everything_becomes_f/tests/test_l2_integration.c
/**
 * test_l2_integration.c - L2 Integration Tests
 * ==============================================
 * 
 * Tests interactions between components.
 * Verifies that components work together correctly.
 */

#include <stdio.h>
#include <string.h>
#include "test_framework.h"
#include "../include/counter.h"
#include "../include/electromagnetic_lock.h"
#include "../include/security_camera.h"
#include "../include/sealed_room.h"
#include "../include/event_system.h"
#include "../include/red_magic_system.h"

/* ============================================================================
 * Counter + Lock Integration Tests
 * ============================================================================ */

static LockSystem *g_test_lock_system = NULL;

static void overflow_callback_trigger_failsafe(UnsignedShortCounter *counter, void *context) {
    (void)counter;
    LockSystem *system = (LockSystem *)context;
    if (system) {
        lock_system_trigger_global_failsafe(system);
    }
}

TEST(test_overflow_triggers_lock_release) {
    UnsignedShortCounter counter;
    LockSystem lock_system;
    
    counter_init(&counter);
    lock_system_init(&lock_system, "Test System");
    lock_system_create_lock(&lock_system, "lock_1", "Door 1");
    lock_system_create_lock(&lock_system, "lock_2", "Door 2");
    
    counter_register_overflow_callback(&counter, overflow_callback_trigger_failsafe, &lock_system);
    
    /* Verify locks are initially locked */
    ASSERT_TRUE(lock_system_all_locked(&lock_system));
    
    /* Cause overflow */
    counter_set_value(&counter, 0xFFFF);
    counter_increment(&counter);
    
    /* Verify locks are now unlocked */
    ASSERT_TRUE(lock_system_all_unlocked(&lock_system));
    ASSERT_TRUE(lock_system_is_failsafe_triggered(&lock_system));
}

TEST(test_fast_forward_to_overflow_unlocks) {
    UnsignedShortCounter counter;
    LockSystem lock_system;
    
    counter_init(&counter);
    lock_system_init(&lock_system, "Test System");
    lock_system_create_lock(&lock_system, "lock_1", "Door 1");
    
    counter_register_overflow_callback(&counter, overflow_callback_trigger_failsafe, &lock_system);
    
    /* Fast forward past overflow */
    counter_fast_forward(&counter, 65536);
    
    ASSERT_TRUE(lock_system_all_unlocked(&lock_system));
}

/* ============================================================================
 * Event Propagation Tests
 * ============================================================================ */

static EventType g_received_events[20];
static int g_event_count = 0;

static void track_event_handler(const Event *event, void *context) {
    (void)context;
    if (g_event_count < 20) {
        g_received_events[g_event_count++] = event->event_type;
    }
}

TEST(test_overflow_event_chain) {
    EventBus bus;
    event_bus_init(&bus, "Test");
    g_event_count = 0;
    
    event_bus_subscribe_all(&bus, track_event_handler, NULL);
    
    /* Simulate event chain */
    event_bus_emit(&bus, EVT_COUNTER_OVERFLOW, "counter", 0);
    event_bus_emit(&bus, EVT_FAILSAFE_TRIGGERED, "system", 0);
    event_bus_emit(&bus, EVT_LOCKS_RELEASED, "locks", 0);
    event_bus_emit(&bus, EVT_ROOM_BREACHED, "room", 0);
    
    ASSERT_EQ(g_event_count, 4);
    ASSERT_EQ(g_received_events[0], EVT_COUNTER_OVERFLOW);
    ASSERT_EQ(g_received_events[1], EVT_FAILSAFE_TRIGGERED);
    ASSERT_EQ(g_received_events[2], EVT_LOCKS_RELEASED);
    ASSERT_EQ(g_received_events[3], EVT_ROOM_BREACHED);
}

static LockSystem *g_chain_lock_system = NULL;
static SealedRoom *g_chain_room = NULL;
static EventBus *g_chain_bus = NULL;

static void on_overflow_trigger_failsafe(const Event *event, void *context) {
    (void)event;
    (void)context;
    if (g_chain_lock_system) {
        lock_system_trigger_global_failsafe(g_chain_lock_system);
        event_bus_emit(g_chain_bus, EVT_LOCKS_RELEASED, "lock_system", 0);
    }
}

static void on_locks_released_breach_room(const Event *event, void *context) {
    (void)event;
    (void)context;
    if (g_chain_room) {
        sealed_room_breach(g_chain_room, 0);
        event_bus_emit(g_chain_bus, EVT_ROOM_BREACHED, "room", 0);
    }
}

TEST(test_event_bus_component_communication) {
    EventBus bus;
    LockSystem lock_system;
    SealedRoom room;
    
    event_bus_init(&bus, "Test");
    lock_system_init(&lock_system, "Test System");
    lock_system_create_lock(&lock_system, "lock_1", "Door 1");
    sealed_room_init(&room, "room_1", "Test Room");
    
    g_chain_lock_system = &lock_system;
    g_chain_room = &room;
    g_chain_bus = &bus;
    
    event_bus_subscribe(&bus, EVT_COUNTER_OVERFLOW, on_overflow_trigger_failsafe, NULL);
    event_bus_subscribe(&bus, EVT_LOCKS_RELEASED, on_locks_released_breach_room, NULL);
    
    /* Trigger the chain */
    event_bus_emit(&bus, EVT_COUNTER_OVERFLOW, "counter", 0);
    
    /* Verify end state */
    ASSERT_TRUE(lock_system_all_unlocked(&lock_system));
    ASSERT_TRUE(sealed_room_is_breached(&room));
    
    g_chain_lock_system = NULL;
    g_chain_room = NULL;
    g_chain_bus = NULL;
}

/* ============================================================================
 * Clock Sync Vulnerability + Camera Tests
 * ============================================================================ */

TEST(test_vulnerability_creates_blind_spot) {
    SecurityCamera cam1, cam2;
    camera_init(&cam1, "cam_1", "Room 1");
    camera_init(&cam2, "cam_2", "Room 2");
    
    /* Record some footage first */
    for (int minute = 0; minute < 60; minute++) {
        camera_record_minute(&cam1, 100, minute);
        camera_record_minute(&cam2, 100, minute);
    }
    
    ClockSyncVulnerability vuln;
    clock_vuln_init(&vuln);
    clock_vuln_add_camera(&vuln, &cam1);
    clock_vuln_add_camera(&vuln, &cam2);
    
    int affected = clock_vuln_exploit(&vuln, 100, 30);
    
    ASSERT_EQ(affected, 2);
    ASSERT_TRUE(camera_has_blind_spots(&cam1));
    ASSERT_TRUE(camera_has_blind_spots(&cam2));
}

TEST(test_reinstall_simulation) {
    SecurityCamera cam;
    camera_init(&cam, "cam_1", "Room 1");
    
    ClockSyncVulnerability vuln;
    clock_vuln_init(&vuln);
    clock_vuln_add_camera(&vuln, &cam);
    
    ReinstallResult result = clock_vuln_simulate_reinstall(&vuln, 50, 15);
    
    ASSERT_TRUE(result.reinstall_triggered);
    ASSERT_TRUE(result.blind_spot_created);
    ASSERT_EQ(result.clock_drift_minutes, 1);
    ASSERT_EQ(result.blind_spot_hour, 50);
    ASSERT_EQ(result.blind_spot_minute, 15);
}

/* ============================================================================
 * Red Magic System Integration Tests
 * ============================================================================ */

TEST(test_system_state_transitions) {
    RedMagicSystem system;
    MagataQuarters quarters;
    
    create_standard_facility(&system, &quarters);
    red_magic_start(&system);
    
    ASSERT_EQ(red_magic_get_state(&system), SYS_STATE_RUNNING);
    
    red_magic_fast_forward_to_overflow(&system);
    
    ASSERT_EQ(red_magic_get_state(&system), SYS_STATE_SYSTEM_DOWN);
}

TEST(test_system_initializing_to_running) {
    RedMagicSystem system;
    red_magic_init(&system, "Test System");
    red_magic_initialize(&system);
    
    ASSERT_EQ(red_magic_get_state(&system), SYS_STATE_INITIALIZING);
    
    red_magic_start(&system);
    
    ASSERT_EQ(red_magic_get_state(&system), SYS_STATE_RUNNING);
}

TEST(test_overflow_triggers_full_sequence) {
    RedMagicSystem system;
    MagataQuarters quarters;
    
    create_standard_facility(&system, &quarters);
    red_magic_start(&system);
    
    red_magic_fast_forward_to_overflow(&system);
    
    ASSERT_EQ(red_magic_get_state(&system), SYS_STATE_SYSTEM_DOWN);
    ASSERT_TRUE(lock_system_is_failsafe_triggered(&system.lock_system));
    ASSERT_TRUE(sealed_room_is_breached(&quarters.base));
}

TEST(test_system_events_during_overflow) {
    RedMagicSystem system;
    MagataQuarters quarters;
    
    create_standard_facility(&system, &quarters);
    g_event_count = 0;
    
    event_bus_subscribe_all(&system.event_bus, track_event_handler, NULL);
    
    red_magic_start(&system);
    red_magic_fast_forward_to_overflow(&system);
    
    /* Check key events occurred */
    bool has_overflow = false, has_failsafe = false, has_locks = false, has_crash = false;
    for (int i = 0; i < g_event_count; i++) {
        if (g_received_events[i] == EVT_COUNTER_OVERFLOW) has_overflow = true;
        if (g_received_events[i] == EVT_FAILSAFE_TRIGGERED) has_failsafe = true;
        if (g_received_events[i] == EVT_LOCKS_RELEASED) has_locks = true;
        if (g_received_events[i] == EVT_SYSTEM_CRASH) has_crash = true;
    }
    
    ASSERT_TRUE(has_overflow);
    ASSERT_TRUE(has_failsafe);
    ASSERT_TRUE(has_locks);
    ASSERT_TRUE(has_crash);
}

TEST(test_standard_facility_setup) {
    RedMagicSystem system;
    MagataQuarters quarters;
    
    create_standard_facility(&system, &quarters);
    
    ASSERT_GTE(lock_system_get_lock_count(&system.lock_system), 1);
    ASSERT_GTE(red_magic_get_camera_count(&system), 1);
    ASSERT_GTE(red_magic_get_sealed_room_count(&system), 1);
    ASSERT_EQ(red_magic_get_state(&system), SYS_STATE_INITIALIZING);
}

/* ============================================================================
 * Room + Lock Integration Tests
 * ============================================================================ */

TEST(test_room_breach_when_locks_released) {
    SealedRoom room;
    ElectromagneticLock lock;
    
    sealed_room_init(&room, "test_room", "Test Room");
    lock_init(&lock, "lock_1", "Main Lock");
    sealed_room_add_door(&room, "door_1", "Main Door", true, &lock);
    
    /* Release the lock */
    lock_trigger_failsafe(&lock);
    
    /* Room should still be sealed (lock release doesn't auto-breach) */
    ASSERT_TRUE(sealed_room_is_sealed(&room));
    
    /* But now breach is possible */
    sealed_room_breach(&room, 0);
    ASSERT_TRUE(sealed_room_is_breached(&room));
}

TEST(test_magata_quarters_full_escape_sequence) {
    MagataQuarters quarters;
    magata_quarters_init(&quarters);
    
    /* Initial state */
    ASSERT_TRUE(sealed_room_is_sealed(&quarters.base));
    ASSERT_TRUE(lock_is_locked(&quarters.main_lock));
    
    /* Trigger overflow escape */
    magata_quarters_trigger_overflow_escape(&quarters, 65536);
    
    /* Final state */
    ASSERT_EQ(sealed_room_get_state(&quarters.base), ROOM_STATE_ESCAPED);
    ASSERT_TRUE(lock_is_unlocked(&quarters.main_lock));
    ASSERT_FALSE(sealed_room_occupant_present(&quarters.base));
}

/* ============================================================================
 * Run L2 Tests
 * ============================================================================ */

int run_l2_tests(void) {
    printf("\n=== L2 Integration Tests ===\n\n");
    
    TEST_SUITE_BEGIN();
    
    /* Counter + Lock integration */
    printf("Counter + Lock Integration:\n");
    RUN_TEST(test_overflow_triggers_lock_release);
    RUN_TEST(test_fast_forward_to_overflow_unlocks);
    
    /* Event propagation */
    printf("\nEvent Propagation:\n");
    RUN_TEST(test_overflow_event_chain);
    RUN_TEST(test_event_bus_component_communication);
    
    /* Camera vulnerability */
    printf("\nClock Sync Vulnerability:\n");
    RUN_TEST(test_vulnerability_creates_blind_spot);
    RUN_TEST(test_reinstall_simulation);
    
    /* Red Magic System */
    printf("\nRed Magic System:\n");
    RUN_TEST(test_system_state_transitions);
    RUN_TEST(test_system_initializing_to_running);
    RUN_TEST(test_overflow_triggers_full_sequence);
    RUN_TEST(test_system_events_during_overflow);
    RUN_TEST(test_standard_facility_setup);
    
    /* Room + Lock */
    printf("\nRoom + Lock Integration:\n");
    RUN_TEST(test_room_breach_when_locks_released);
    RUN_TEST(test_magata_quarters_full_escape_sequence);
    
    TEST_SUMMARY();
    
    return TEST_GET_FAILED();
}
